File: src/main/services/configParser.ts
import yaml from 'js-yaml';

type ConfigFormat = 'yaml' | 'json' | 'properties';

export interface KeyValueMatch {
  keyPath: string;
  value: unknown;
}

export function findKeyValue(content: string, keyPath: string, dataId: string, type?: string): KeyValueMatch | null {
  const format = detectConfigFormat(dataId, type);

  if (format === 'properties') {
    const properties = parseProperties(content);
    return properties.has(keyPath) ? { keyPath, value: properties.get(keyPath) } : null;
  }

  const parsed = parseStructuredContent(content, format);
  const value = getNestedValue(parsed, keyPath);

  return value === undefined ? null : { keyPath, value };
}

export function upsertKeyValue(
  content: string,
  keyPath: string,
  value: unknown,
  dataId: string,
  type?: string
): string {
  const format = detectConfigFormat(dataId, type);

  if (format === 'properties') {
    return upsertPropertiesValue(content, keyPath, value);
  }

  const parsed = parseStructuredContent(content, format);
  const next = isRecord(parsed) ? parsed : {};
  setNestedValue(next, keyPath, value);

  if (format === 'json') {
    return `${JSON.stringify(next, null, 2)}\n`;
  }

  return yaml.dump(next, { lineWidth: -1 });
}

export function detectConfigFormat(dataId: string, type?: string): ConfigFormat {
  const normalizedType = type?.toLowerCase();
  const normalizedDataId = dataId.toLowerCase();

  if (normalizedType === 'yaml' || normalizedType === 'yml' || normalizedDataId.endsWith('.yml') || normalizedDataId.endsWith('.yaml')) {
    return 'yaml';
  }

  if (normalizedType === 'json' || normalizedDataId.endsWith('.json')) {
    return 'json';
  }

  return 'properties';
}

function parseStructuredContent(content: string, format: Exclude<ConfigFormat, 'properties'>): unknown {
  if (!content.trim()) {
    return {};
  }

  return format === 'json' ? JSON.parse(content) : yaml.load(content);
}

function getNestedValue(input: unknown, keyPath: string): unknown {
  return keyPath.split('.').reduce<unknown>((current, segment) => {
    if (!isRecord(current)) {
      return undefined;
    }

    return current[segment];
  }, input);
}

function setNestedValue(input: Record<string, unknown>, keyPath: string, value: unknown): void {
  const segments = keyPath.split('.');
  let current = input;

  for (const [index, segment] of segments.entries()) {
    if (index === segments.length - 1) {
      current[segment] = value;
      return;
    }

    if (!isRecord(current[segment])) {
      current[segment] = {};
    }

    current = current[segment] as Record<string, unknown>;
  }
}

function parseProperties(content: string): Map<string, string> {
  const properties = new Map<string, string>();

  for (const line of content.split(/\r?\n/)) {
    const trimmed = line.trim();

    if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('!')) {
      continue;
    }

    const separatorIndex = findPropertySeparatorIndex(line);
    if (separatorIndex === -1) {
      continue;
    }

    const key = line.slice(0, separatorIndex).trim();
    const value = line.slice(separatorIndex + 1).trim();
    properties.set(key, value);
  }

  return properties;
}

function upsertPropertiesValue(content: string, keyPath: string, value: unknown): string {
  const lines = content.split(/\r?\n/);
  const stringValue = String(value);
  let updated = false;

  const nextLines = lines.map((line) => {
    const separatorIndex = findPropertySeparatorIndex(line);

    if (separatorIndex === -1 || line.trim().startsWith('#') || line.trim().startsWith('!')) {
      return line;
    }

    const key = line.slice(0, separatorIndex).trim();
    if (key !== keyPath) {
      return line;
    }

    updated = true;
    return `${keyPath}=${stringValue}`;
  });

  if (!updated) {
    if (nextLines.length > 0 && nextLines[nextLines.length - 1] !== '') {
      nextLines.push(`${keyPath}=${stringValue}`);
    } else {
      nextLines[nextLines.length - 1] = `${keyPath}=${stringValue}`;
    }
  }

  return `${nextLines.join('\n')}\n`;
}

function findPropertySeparatorIndex(line: string): number {
  const equalIndex = line.indexOf('=');
  const colonIndex = line.indexOf(':');

  if (equalIndex === -1) {
    return colonIndex;
  }

  if (colonIndex === -1) {
    return equalIndex;
  }

  return Math.min(equalIndex, colonIndex);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value));
}

File: native/windows/src/NacosSyncTool.Windows/Services/ConfigParser.cs
using System.Text.Json;
using System.Collections;
using System.Globalization;
using NacosSyncTool.Windows.Models;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace NacosSyncTool.Windows.Services;

/// <summary>
/// 配置格式
/// </summary>
public enum ConfigFormat
{
    Yaml,
    Json,
    Properties
}

/// <summary>
/// Key 值匹配结果
/// </summary>
public class KeyValueMatch
{
    public string KeyPath { get; set; } = string.Empty;
    public object? Value { get; set; }
}

/// <summary>
/// 配置解析器
/// 支持 YAML / JSON / Properties 三种格式
/// </summary>
public static class ConfigParser
{
    private static readonly IDeserializer YamlDeserializer = new DeserializerBuilder()
        .WithNamingConvention(NullNamingConvention.Instance)
        .Build();

    private static readonly ISerializer YamlSerializer = new SerializerBuilder()
        .WithNamingConvention(NullNamingConvention.Instance)
        .Build();

    /// <summary>
    /// 检测配置格式
    /// </summary>
    public static ConfigFormat DetectConfigFormat(string dataId, string? type)
    {
        var normalizedType = type?.Trim().ToLowerInvariant();
        var normalizedDataId = dataId.ToLowerInvariant();

        if (normalizedType == "yaml" || normalizedType == "yml"
            || normalizedDataId.EndsWith(".yml") || normalizedDataId.EndsWith(".yaml"))
        {
            return ConfigFormat.Yaml;
        }

        if (normalizedType == "json" || normalizedDataId.EndsWith(".json"))
        {
            return ConfigFormat.Json;
        }

        return ConfigFormat.Properties;
    }

    /// <summary>
    /// 在配置内容中查找指定 Key
    /// </summary>
    public static KeyValueMatch? FindKeyValue(string content, string keyPath, string dataId, string? type)
    {
        var format = DetectConfigFormat(dataId, type);

        if (format == ConfigFormat.Properties)
        {
            var properties = ParseProperties(content);
            return properties.TryGetValue(keyPath, out var propValue)
                ? new KeyValueMatch { KeyPath = keyPath, Value = propValue }
                : null;
        }

        var parsed = ParseStructuredContent(content, format);
        var nestedValue = GetNestedValue(parsed, keyPath);

        return nestedValue == null ? null : new KeyValueMatch { KeyPath = keyPath, Value = nestedValue };
    }

    /// <summary>
    /// 插入或更新指定 Key 的值
    /// </summary>
    public static string UpsertKeyValue(string content, string keyPath, object? value, string dataId, string? type)
    {
        var format = DetectConfigFormat(dataId, type);

        if (format == ConfigFormat.Properties)
        {
            return UpsertPropertiesValue(content, keyPath, value);
        }

        var parsed = ParseStructuredContent(content, format);
        var dict = parsed as Dictionary<string, object?> ?? new Dictionary<string, object?>();
        SetNestedValue(dict, keyPath, value);

        if (format == ConfigFormat.Json)
        {
            return JsonSerializer.Serialize(dict, new JsonSerializerOptions { WriteIndented = true }) + "\n";
        }

        return YamlSerializer.Serialize(dict);
    }

    /// <summary>
    /// 解析结构化内容（YAML/JSON）
    /// </summary>
    private static object? ParseStructuredContent(string content, ConfigFormat format)
    {
        if (string.IsNullOrWhiteSpace(content))
        {
            return new Dictionary<string, object?>();
        }

        try
        {
            if (format == ConfigFormat.Json)
            {
                using var document = JsonDocument.Parse(content);
                return NormalizeStructuredValue(document.RootElement);
            }

            return NormalizeStructuredValue(YamlDeserializer.Deserialize<object>(content));
        }
        catch
        {
            return new Dictionary<string, object?>();
        }
    }

    private static object? NormalizeStructuredValue(object? value)
    {
        if (value is null || value is string || IsScalar(value))
        {
            return value;
        }

        if (value is JsonElement jsonElement)
        {
            return NormalizeJsonElement(jsonElement);
        }

        if (value is IDictionary dictionary)
        {
            var result = new Dictionary<string, object?>();
            foreach (DictionaryEntry entry in dictionary)
            {
                var key = Convert.ToString(entry.Key, CultureInfo.InvariantCulture);
                if (!string.IsNullOrEmpty(key))
                {
                    result[key] = NormalizeStructuredValue(entry.Value);
                }
            }
            return result;
        }

        if (value is IEnumerable enumerable)
        {
            return enumerable.Cast<object?>().Select(NormalizeStructuredValue).ToList();
        }

        return value;
    }

    private static object? NormalizeJsonElement(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.Object => element.EnumerateObject()
                .ToDictionary(property => property.Name, property => NormalizeJsonElement(property.Value)),
            JsonValueKind.Array => element.EnumerateArray().Select(NormalizeJsonElement).ToList(),
            JsonValueKind.String => element.GetString(),
            JsonValueKind.Number => element.TryGetInt64(out var longValue)
                ? longValue
                : element.TryGetDouble(out var doubleValue) ? doubleValue : element.GetRawText(),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            _ => element.GetRawText()
        };
    }

    private static bool IsScalar(object value)
    {
        var type = value.GetType();
        return type.IsPrimitive
               || value is decimal
               || value is DateTime
               || value is DateTimeOffset
               || value is Guid
               || value is Enum;
    }

    /// <summary>
    /// 获取嵌套值（支持 a.b.c 路径）
    /// </summary>
    private static object? GetNestedValue(object? input, string keyPath)
    {
        var current = input;
        foreach (var segment in keyPath.Split('.'))
        {
            if (current is not Dictionary<string, object?> dict || !dict.TryGetValue(segment, out var value))
            {
                return null;
            }
            current = value;
        }

        return current;
    }

    /// <summary>
    /// 设置嵌套值
    /// </summary>
    private static void SetNestedValue(Dictionary<string, object?> input, string keyPath, object? value)
    {
        var segments = keyPath.Split('.');
        var current = input;

        for (int i = 0; i < segments.Length; i++)
        {
            var segment = segments[i];

            if (i == segments.Length - 1)
            {
                current[segment] = value;
                return;
            }

            if (!current.TryGetValue(segment, out var existing)
                || existing is not Dictionary<string, object?>)
            {
                var nested = new Dictionary<string, object?>();
                current[segment] = nested;
                current = nested;
            }
            else
            {
                current = (Dictionary<string, object?>)existing;
            }
        }
    }

    /// <summary>
    /// 解析 Properties 格式
    /// </summary>
    private static Dictionary<string, string> ParseProperties(string content)
    {
        var properties = new Dictionary<string, string>();

        foreach (var line in content.Split('\n', '\r'))
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#') || trimmed.StartsWith('!'))
            {
                continue;
            }

            var separatorIndex = FindPropertySeparatorIndex(line);
            if (separatorIndex == -1)
            {
                continue;
            }

            var key = line.Substring(0, separatorIndex).Trim();
            var value = line.Substring(separatorIndex + 1).Trim();
            properties[key] = value;
        }

        return properties;
    }

    /// <summary>
    /// 插入或更新 Properties 值
    /// </summary>
    private static string UpsertPropertiesValue(string content, string keyPath, object? value)
    {
        var lines = content.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None).ToList();
        var stringValue = ConfigValueFormatter.ToDisplayText(value).Replace("\r\n", " ").Replace('\n', ' ');
        bool updated = false;

        for (int i = 0; i < lines.Count; i++)
        {
            var line = lines[i];
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#') || trimmed.StartsWith('!'))
            {
                continue;
            }

            var separatorIndex = FindPropertySeparatorIndex(line);
            if (separatorIndex == -1)
            {
                continue;
            }

            var key = line.Substring(0, separatorIndex).Trim();
            if (key == keyPath)
            {
                lines[i] = $"{keyPath}={stringValue}";
                updated = true;
            }
        }

        if (!updated)
        {
            if (lines.Count > 0 && lines[lines.Count - 1] != "")
            {
                lines.Add($"{keyPath}={stringValue}");
            }
            else if (lines.Count > 0)
            {
                lines[lines.Count - 1] = $"{keyPath}={stringValue}";
            }
            else
            {
                lines.Add($"{keyPath}={stringValue}");
            }
        }

        return string.Join("\n", lines) + "\n";
    }

    /// <summary>
    /// 查找 Properties 分隔符位置（= 或 :）
    /// </summary>
    private static int FindPropertySeparatorIndex(string line)
    {
        int equalIndex = line.IndexOf('=');
        int colonIndex = line.IndexOf(':');

        if (equalIndex == -1) return colonIndex;
        if (colonIndex == -1) return equalIndex;
        return Math.Min(equalIndex, colonIndex);
    }
}
